hoptimator-kafka-adapter/src/main/java/com/linkedin/hoptimator/operator/kafka/KafkaTopicAclReconciler.java

```diff
@@ -0,0 +1,134 @@
+import org.apache.kafka.common.resource.PatternType;
+import org.apache.kafka.common.resource.ResourcePattern;
+import org.apache.kafka.common.resource.ResourceType;
+import org.apache.kafka.clients.admin.AdminClient;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Properties;
+import java.util.concurrent.ExecutionException;
+
+public class KafkaTopicAclReconciler implements Reconciler {
+  private final static Logger log = LoggerFactory.getLogger(KafkaTopicAclReconciler.class);
+  private final static String ACL = "hoptimator.linkedin.com/v1alpha1/Acl";
+  private final static String KAFKATOPIC = "hoptimator.linkedin.com/v1alpha1/KafkaTopic";
+
+  private final Operator operator;
+
+  public KafkaTopicAclReconciler(Operator operator) {
+    this.operator = operator;
+  }
+
+  @Override
+  public Result reconcile(Request request) {
+    log.info("Reconciling request {}", request);
+    String name = request.getName();
+    String namespace = request.getNamespace();
+
+    try {
+      V1alpha1Acl object = operator.<V1alpha1Acl>fetch(ACL, namespace, name);
+
+      if (object == null) {
+        log.info("Object {}/{} deleted. Skipping.", namespace, name);
+        return new Result(false);
+      }
+
+      String targetKind = object.getSpec().getResource().getKind();
+
+      if (!targetKind.equals("KafkaTopic")) {
+        log.info("Not a KafkaTopic Acl. Skipping.");
+      }
+
+      V1alpha1AclSpec.MethodEnum method = object.getSpec().getMethod();
+      AclOperation operation;
+      switch (method) {
+      case READ:
+        operation = AclOperation.READ;
+        break;
+      case WRITE:
+        operation = AclOperation.WRITE;
+        break;
+      default:
+        log.info("Unsupported KafkaTopic Acl operation {}. Skipping.", method);
+        return new Result(false);
+      }
+      
+      String targetName = object.getSpec().getResource().getName();
+      String principal = object.getSpec().getPrincipal();
+
+      V1alpha1KafkaTopic target = operator.<V1alpha1KafkaTopic>fetch(KAFKATOPIC, namespace, targetName);
+
+      if (target == null) {
+        log.info("Target KafkaTopic {}/{} not found. Retrying.", namespace, targetName);
+        return new Result(true, operator.failureRetryDuration());
+      }
+
+      // assemble AdminClient config
+      ConfigAssembler assembler = new ConfigAssembler(operator);
+      list(target.getSpec().getClientConfigs()).forEach(x ->
+        assembler.addRef(namespace, x.getConfigMapRef().getName()));
+      map(target.getSpec().getClientOverrides()).forEach((k, v) -> assembler.addOverride(k, v));
+      Properties properties = assembler.assembleProperties();
+      log.info("Using AdminClient config: {}", properties);
+
+      AdminClient admin = AdminClient.create(properties);
```

Comment:
Isn't this an expensive operation? Especially if the reconciliation keeps on failing for some other internal reason.

Reply:
We have a non-zero retry duration, so reconciliation shouldn't be a tight loop. But yeah, it can be sorta expensive, since it registers a bunch of metrics and whatnot.